src/context/Baser.ts

```diff
@@ -0,0 +1,174 @@
+		try {
+			await workspace.openTextDocument(Uri.parse(path))
+		} catch (error) {
+			return false
+		}
+		return true
+	}
+
+	private activeInfoMessages = new Set<string>() // Prevent repeat message animations when arrowing through many results with the same uri.
+	public async translateArtifactToLocal(artifactPath: string) { // Retval is validated.
+		// Temp.
+		if (artifactPath.startsWith('sarif:')) return artifactPath
+		const validateUri = async () => {
+			// Cache
+			if (this.validatedPathsArtifactToLocal.has(artifactPath))
+				return this.validatedPathsArtifactToLocal.get(artifactPath)
+
+			// File System Exist
+			if (await this.pathExists(artifactPath))
+				return artifactPath
+
+			// Known Bases
+			for (const [artifactBase, localBase] of this.basesArtifactToLocal) {
+				if (!artifactPath.startsWith(artifactBase)) continue // Just let it fall through?
+
+				const normalizedArtifactPath = `${artifactPath.startsWith('/') ? '' : '/'}${artifactPath}`
+				const localPath = normalizedArtifactPath.replace(artifactBase, localBase)
+
+				if (await this.pathExists(localPath)) {
+					this.updateValidatedPaths(artifactPath, localPath)
+					return localPath
+				}
+			}
+
+			{ // API-injected baseUris
+				const localPath = await this.tryUriBases(artifactPath)
+				if (localPath) return localPath
+			}
+
+			// Distinct Project Items
+			const {file} = artifactPath
+			if (this.distinctLocalNames.has(file) && this.store.distinctArtifactNames.has(file)) {
+				const localPath = this.distinctLocalNames.get(file)
+				this.updateValidatedPaths(artifactPath, localPath)
+				this.updateBases(artifactPath.split('/'), localPath.split('/'))
+				return localPath
+			}
+
+			// Open Docs
+			for (const doc of workspace.textDocuments) {
+				const localPath = doc.uri.path
+				if (localPath.file !== artifactPath.file) continue
+				this.updateValidatedPaths(artifactPath, localPath)
+				this.updateBases(artifactPath.split('/'), localPath.split('/'))
+				return localPath
+			}
+
+			return '' // Can't find uri.
+		}
+
+		let validatedUri = await validateUri()
+		if (!validatedUri && !this.activeInfoMessages.has(artifactPath)) {
+			this.activeInfoMessages.add(artifactPath)
+			const choice = await window.showInformationMessage(`Unable to find '${artifactPath.split('/').pop()}'`, 'Locate...')
+			this.activeInfoMessages.delete(artifactPath)
+			if (choice === 'Locate...') {
+				const extension = artifactPath.match(/\.([\w]+)$/)[1]
+				const files = await window.showOpenDialog({
+					defaultUri: workspace.workspaceFolders?.[0]?.uri,
+					filters: { 'Matching file' : [extension] },
+					// Consider allowing folders.
+				})
+				if (!files?.length) return // User cancelled.
+
+				const partsOld = artifactPath.split('/')
+				const partsNew = files[0]?.path.split('/')
+				if (partsOld.last !== partsNew.last) {
+					void window.showErrorMessage(`File names must match: "${partsOld.last}" and "${partsNew.last}"`)
+					return
```

Comment:
offer to try again?

Reply:
I think that makes sense.

In the larger picture, the "Locate" dialog (although less intrusive than the previous design) is still too intrusive and noisy. The next design to be trialed is to have the "Locate..." button in the details pane. If that is implemented, then it would be relatively easy to "Locate" again.

Reply:
I think that makes sense.

In the larger picture, the "Locate" dialog (although less intrusive than the previous design) is still too intrusive and noisy. The next design to be trialed is to have the "Locate..." button in the details pane. If that is implemented, then it would be relatively easy to "Locate" again.

---
In reply to: [430751162](https://github.com/microsoft/sarif-vscode-extension/pull/279#discussion_r430751162) [](ancestors = 430751162)